core/src/main/java/com/yahoo/oak/SliceSyncRecycle.java

```diff
@@ -0,0 +1,250 @@
+// which is the aforementioned portion of an off-heap memory.
+class SliceSyncRecycle extends Slice {
+    static final int UNDEFINED_LENGTH_OR_OFFSET = -1;
+
+    /**
+     * An allocated by SyncRecycleMemoryManager off-heap cut have reserved space for meta-data, i.e., a header.
+     * The header size is defined externally by memory-manager at the slice construction.
+     */
+    private final int headerSize;
+    private final SyncRecycleMMHeader header;
+
+    private final long invalidReferenceValue; // used for invalidation
+
+
+    /* ------------------------------------------------------------------------------------
+     * Constructors
+     * ------------------------------------------------------------------------------------*/
+    // Should be used only by Memory Manager (within Memory Manager package)
+    SliceSyncRecycle(int headerSize, SyncRecycleMMHeader header) {
+        super();
+        this.headerSize = headerSize;
+        this.invalidReferenceValue = ReferenceCodecSyncRecycle.getInvalidReference();
+        this.header = header;
+        invalidate();
+    }
+
+    // Should be used only for testing
+    @VisibleForTesting SliceSyncRecycle() {
+        this(0, null);
+    }
+
+    // Used to duplicate the allocation state. Does not duplicate the underlying memory buffer itself.
+    // Should be used when ThreadContext's internal Slice needs to be exported to the user.
+    SliceSyncRecycle getDuplicatedSlice() {
+        SliceSyncRecycle newSlice = new SliceSyncRecycle(this.headerSize, this.header);
+        newSlice.copyFrom(this);
+        return newSlice;
+    }
+
+    /* ------------------------------------------------------------------------------------
+     * Allocation info and metadata setters
+     * ------------------------------------------------------------------------------------*/
+    // Reset all not final fields to invalid state
+    void invalidate() {
+        blockID     = NativeMemoryAllocator.INVALID_BLOCK_ID;
+        reference   = invalidReferenceValue;
+        length      = UNDEFINED_LENGTH_OR_OFFSET;
+        offset      = UNDEFINED_LENGTH_OR_OFFSET;
+        buffer      = null;
+        associated  = false;
+    }
+
+    /*
+     * Updates everything that can be extracted with reference decoding of "SyncRecycle" type,
+     * including reference itself. The separation by reference decoding type is temporary!
+     * This is not the full setting of the association, therefore 'associated' flag remains false
+     */
+    void associateReferenceDecoding(int blockID, int offset, int version, long reference) {
+        setBlockIdOffsetAndLength(blockID, offset, UNDEFINED_LENGTH_OR_OFFSET);
+        this.reference = reference;
+        this.version = version;
+        associated   = false;
+    }
+
+    // Copy the block allocation information from another block allocation.
+    <T extends Slice> void copyFrom(T other) {
+        if (other instanceof SliceSyncRecycle) { //TODO: any other ideas?
```

Comment:
```suggestion
        if (!(other instanceof SliceSyncRecycle)) throw...
        
        this.copyAllocationInfoFrom(other);
        // Init other fields.
```

Reply:
I thought maybe somehow to avoid `instanceof`

Reply:
This will be resolved once we transfer majority of the interface from` MemoryManager` to `Slice`.